build-tools/packages/build-infrastructure/src/workspace.ts

```diff
@@ -0,0 +1,198 @@
+/*!
+ * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+import path from "node:path";
+
+import { getPackagesSync } from "@manypkg/get-packages";
+import execa from "execa";
+
+import type { ReleaseGroupDefinition, WorkspaceDefinition } from "./config.js";
+import { loadPackageFromWorkspaceDefinition } from "./package.js";
+import { createPackageManager } from "./packageManagers.js";
+import { ReleaseGroup } from "./releaseGroup.js";
+import type {
+	IPackage,
+	IPackageManager,
+	IReleaseGroup,
+	IWorkspace,
+	ReleaseGroupName,
+	WorkspaceName,
+} from "./types.js";
+
+/**
+ * {@inheritDoc IWorkspace}
+ */
+export class Workspace implements IWorkspace {
+	/**
+	 * {@inheritDoc IWorkspace.name}
+	 */
+	public readonly name: WorkspaceName;
+
+	/**
+	 * {@inheritDoc IWorkspace.releaseGroups}
+	 */
+	public readonly releaseGroups: Map<ReleaseGroupName, IReleaseGroup>;
+
+	/**
+	 * {@inheritDoc IWorkspace.rootPackage}
+	 */
+	public readonly rootPackage: IPackage;
+
+	/**
+	 * {@inheritDoc IWorkspace.packages}
+	 */
+	public readonly packages: IPackage[];
+
+	/**
+	 * {@inheritDoc IWorkspace.directory}
+	 */
+	public readonly directory: string;
+
+	private readonly packageManager: IPackageManager;
+
+	/**
+	 * Construct a new workspace object.
+	 *
+	 * @param name - The name of the workspace.
+	 * @param definition - The definition of the workspace.
+	 * @param root - The path to the root of the workspace.
+	 */
+	private constructor(name: string, definition: WorkspaceDefinition, root: string) {
+		this.name = name as WorkspaceName;
+		this.directory = path.resolve(root, definition.directory);
+
+		const {
+			tool,
+			packages: foundPackages,
+			rootPackage: foundRootPackage,
+			rootDir: foundRoot,
+		} = getPackagesSync(this.directory);
+		if (foundRoot !== this.directory) {
+			// This is a sanity check. directory is the path passed in when creating the Workspace object, while rootDir is
+			// the dir that manypkg found. They should be the same.
+			throw new Error(
+				`The root dir found by manypkg, '${foundRoot}', does not match the configured directory '${this.directory}'`,
+			);
+		}
+
+		if (foundRootPackage === undefined) {
+			throw new Error(`No root package found for workspace in '${foundRoot}'`);
+		}
+
+		switch (tool.type) {
+			case "npm":
+			case "pnpm":
+			case "yarn": {
+				this.packageManager = createPackageManager(tool.type);
+				break;
+			}
+			default: {
+				throw new Error(`Unknown package manager ${tool.type}`);
+			}
+		}
+
+		this.packages = [];
+		for (const pkg of foundPackages) {
+			const loadedPackage = loadPackageFromWorkspaceDefinition(
+				path.join(pkg.dir, "package.json"),
+				this.packageManager,
+				/* isWorkspaceRoot */ foundPackages.length === 1,
```

Comment:
Hmmmm, it's definitely true for npm, since npm's workspace settings are in package.json itself, so if there's only a
single package, it's the root package of the workspace. I did some testing with pnpm, and it seems to allow the folder
with the pnpm-workspace.yaml file to exist without a package, so technically there could be no "root package" for pnpm.
That said, I don't think that's common enough to support.